File: 08/py/solution.py
import sys

def parseInput(fName):
    with open(fName, 'r') as f:
        input = f.readlines()

    input = [l.rstrip() for l in input]
    return input



def solvePart1(input):
    nRows = len(input)
    nCols = len(input[0])
    # map of trees which are visible
    # 1 == visible, 0 == not visible
    visMap = [[0 for i in range(nCols)] for j in range(nRows)]
    
    # set the edges visible
    for ir in range(nRows):
        visMap[ir][0] = 1
        visMap[ir][nCols-1] = 1
    for ic in range(nCols):
        visMap[0][ic] = 1
        visMap[nRows-1][ic] = 1

    # check left<->right first
    for ir in range(1, nRows-1):
        # visibility from left
        lTallest = input[ir][0]
        for ic in range(1, nCols):
            if input[ir][ic] > lTallest:
                visMap[ir][ic] = 1
                lTallest = input[ir][ic]
        # visibility from right
        rTallest = input[ir][nCols-1]
        for ic in range(nCols - 2, 0, -1):
            if input[ir][ic] > rTallest:
                visMap[ir][ic] = 1
                rTallest = input[ir][ic]

    for ic in range(1, nCols-1):
        # visibility from top
        tTallest = input[0][ic]
        for ir in range (1, nRows):
            if input[ir][ic] > tTallest:
                visMap[ir][ic] = 1
                tTallest = input[ir][ic]
        # visibility from bottom
        bTallest = input[nRows - 1][ic]
        for ir in range(nRows - 2, 0, -1):
            if input[ir][ic] > bTallest:
                visMap[ir][ic] = 1
                bTallest = input[ir][ic]

    totalVis = 0
    for ir in range(nRows):
        for ic in range(nCols):
            if visMap[ir][ic] == 1:
                totalVis += 1
    return totalVis

def solvePart2(input):
    nRows = len(input)
    nCols = len(input[0])
    scenicScores = [[0 for i in range(nCols)] for j in range(nRows)]

    def getVisibility(r, c):
        lVis = 0
        rVis = 0
        uVis = 0
        dVis = 0

        # left
        for ic in range(c-1, -1, -1):
            lVis += 1
            if input[r][ic] >= input[r][c]:
                break

        # right
        for ic in range(c+1, nCols):
            rVis += 1
            if input[r][ic] >= input[r][c]:
                break

        # up
        for ir in range(r+1, nRows):
            uVis += 1
            if input[ir][c] >= input[r][c]:
                break

        # down
        for ir in range (r-1, -1, -1):
            dVis += 1
            if input[ir][c] >= input[r][c]:
                break

        return lVis * rVis * uVis * dVis

    maxScore = 0
    for ir in range(nRows):
        for ic in range(nCols):
            scenicScores[ir][ic] = getVisibility(ir, ic)
            if scenicScores[ir][ic] > maxScore:
                maxScore = scenicScores[ir][ic]
    
    return maxScore

input = parseInput(sys.argv[1])
part1soln = solvePart1(input)
print(f"Part 1 solution is: {part1soln}")
part2soln = solvePart2(input)
print(f"Part 2 solution is: {part2soln}")

File: 08/cpp/solution.cpp
#include<fstream>
#include<iostream>
#include<vector>
#include<string>

using namespace std;
typedef vector<vector<int>> tmap;

vector<string> parseInput(string fName) {
    vector<string> treeMap;
    string line;
    fstream file = fstream(fName);
    if (file.is_open()) {
        while(getline(file, line)) {
            treeMap.push_back(line);
        }
        file.close();
    }

    return treeMap;
}

void checkVisibility(int r, int c, vector<string>& treeMap, tmap& vMap) {
    int nRow = treeMap.size();
    int nCol = treeMap[0].size();

    if (vMap[r][c] == 1 || vMap[r][c] == 0) {
        return;
    }

    // check each direction until we find visibility
    // left
    for (int i=(c-1); i >= 0; i--) {
        if (vMap[r][i] == -1) { checkVisibility(r, i, treeMap, vMap); }

        if (vMap[r][i] == 1) { //neighbor is visible
            if (treeMap[r][c] > treeMap[r][i]) {
                vMap[r][c] = 1;
                return;
            } else {
                vMap[r][c] = 0;
                break;
            }
        } else if (vMap[r][i] == 0) {
            if (treeMap[r][c] < treeMap[r][i]) { 
                vMap[r][c] = 0;
                break; 
            }
        } 
    }
    // right
    for (int i=(c+1); i < nCol; i++) {
        if (vMap[r][i] == -1) { checkVisibility(r, i, treeMap, vMap); }

        if (vMap[r][i] == 1) { //neighbor is visible
            if (treeMap[r][c] > treeMap[r][i]) {
                vMap[r][c] = 1;
                return;
            } else {
                vMap[r][c] = 0;
                break;
            }
        } else if (vMap[r][i] == 0) {
            if (treeMap[r][c] < treeMap[r][i]) { 
                vMap[r][c] = 0;
                break; 
            }
        } 
    }
    // up
    for (int i=(r+1); i < nRow; i++) {
        if (vMap[i][c] == -1) { checkVisibility(i, c, treeMap, vMap); }

        if (vMap[i][c] == 1) { //neighbor is visible
            if (treeMap[r][c] > treeMap[i][c]) {
                vMap[r][c] = 1;
                return;
            } else {
                vMap[r][c] = 0;
                break;
            }
        } else if (vMap[i][c] == 0) {
            if (treeMap[r][c] < treeMap[i][c]) { 
                vMap[r][c] = 0;
                break; 
            }
        }
    }
    // down
    for (int i=(r-1); i >= 0; i--) {
        if (vMap[i][c] == -1) { checkVisibility(i, c, treeMap, vMap); }

        if (vMap[i][c] == 1) { //neighbor is visible
            if (treeMap[r][c] > treeMap[i][c]) {
                vMap[r][c] = 1;
                return;
            } else {
                vMap[r][c] = 0;
                break;
            }
        } else if (vMap[i][c] == 0) {
            if (treeMap[r][c] < treeMap[i][c]) { 
                vMap[r][c] = 0;
                break; 
            }
        }
    }
}

int solvePart1(vector<string> treeMap) {
    int nVisible = 0;
    int nRow = treeMap.size();
    int nCol = treeMap[0].size();

    vector<vector<int>> vMap(
        nRow,
        vector<int>(nCol, -1)
    );

    // fill in the border trees (which are always visible)
    for(int i=0; i<nRow; i++){
        vMap[i][0] = 1;
        vMap[i][nCol - 1] = 1;
    }
    for(int j=0; j<nCol; j++){
        vMap[0][j] = 1;
        vMap[nRow - 1][j] = 1;
    }

    for(int i=1; i<nRow; i++) {
        for (int j=0; j<nCol; j++) {
            checkVisibility(i, j, treeMap, vMap);
        }
    }

    for (auto& row : vMap) {
        for (auto& elem : row) {
            nVisible += elem;
        }
    }

    return nVisible;
}

int main(int argc, char* argv[]) {
    vector<string> input = parseInput(argv[1]);
    int part1soln = solvePart1(input);
    cout << "Part 1 solution is: " << part1soln << "\n";
}